include/envoy/upstream/thread_local_cluster.h

```diff
@@ -8,6 +8,31 @@
 namespace Envoy {
 namespace Upstream {
 
+// HttpPoolData returns information about a given pool as well as a function
+// to create streams on that pool.
+struct HttpPoolData {
+  using StreamCreationFunction = std::function<Envoy::Http::ConnectionPool::Cancellable*(
+      Http::ResponseDecoder& response_decoder, Envoy::Http::ConnectionPool::Callbacks& callbacks)>;
+  // Rather than returning the pool, httpConnPool supplies function to create a stream on the
+  // pool. This allows the ThreadLocalCluster to have its own logic, such as preconnect, ahead of
+  // stream creation.
+  StreamCreationFunction create_stream_;
```

Comment:
It is clear what the lifetime guarantees are for the things which might be captured in this function